# Кан Влдаислав
## AIN-2-22

#### 2)	Создайте свой «мешок слов», например, набор из 20-25 отзывов о чем-либо, напишите краткое пояснение к данным.

In [4]:
reviews = [
    "Фильм был потрясающим, отличная актерская игра.",
    "Мне не понравился фильм, слишком затянутый сюжет.",
    "Очень интересный фильм, всем рекомендую!",
    "Фильм средний, ожидал большего.",
    "Не лучший фильм, но и не худший.",
    "Фантастический фильм! Обязательно к просмотру.",
    "Фильм оставил смешанные чувства.",
    "Сюжет предсказуемый, но все равно хороший фильм.",
    "Фильм удивил, очень понравился.",
    "Скучный фильм, не советую.",
    "Замечательная работа режиссера, фильм захватывает с первых минут.",
    "Диалоги в фильме были слишком затянутыми, местами скучновато.",
    "Яркие спецэффекты, но слабый сюжет.",
    "Фильм вызывает ностальгию, советую всем.",
    "Слишком много клише, но смотрится легко.",
    "Отличный фильм для семейного просмотра.",
    "Актерская игра была на высоте, особенно главного героя.",
    "Неожиданный поворот сюжета в конце, понравилось.",
    "Фильм слишком депрессивный, ожидал чего-то более легкого.",
    "Захватывающий сюжет, невозможно оторваться.",
    "Музыкальное сопровождение идеально дополняло атмосферу.",
    "Фильм оставил равнодушным, ничем не запомнился.",
    "Очень трогательная история, заставила задуматься.",
    "Понравились костюмы и декорации, выглядит великолепно.",
    "Фильм был слишком длинным, можно было сократить на час.",
    "Смешно и трогательно одновременно, понравится всем.",
    "Яркий и вдохновляющий фильм, обязательно к просмотру."
]

##### это набор отзывов о фильме, собранных из различных источников. Отзывы включают в себя как положительные, так и отрицательные мнения, что позволяет получить более полный обзор восприятия фильма аудиторией.

#### 3)	Создайте словарь стоп-слов, примените стоп-слова к своему «мешку слов»

In [6]:
import nltk
from nltk.corpus import stopwords
try:
    stop_words = stopwords.words('russian')
except LookupError:
    nltk.download('stopwords')
    stop_words = stopwords.words('russian')

stop_words = set(stop_words)

print(vectorizer.get_feature_names_out())

['актерская' 'атмосферу' 'большего' 'вдохновляющий' 'великолепно' 'всем'
 'выглядит' 'вызывает' 'высоте' 'героя' 'главного' 'декорации'
 'депрессивный' 'диалоги' 'длинным' 'дополняло' 'задуматься'
 'замечательная' 'запомнился' 'заставила' 'затянутый' 'затянутыми'
 'захватывает' 'захватывающий' 'игра' 'идеально' 'интересный' 'история'
 'клише' 'конце' 'костюмы' 'легко' 'легкого' 'лучший' 'местами' 'минут'
 'музыкальное' 'невозможно' 'неожиданный' 'ничем' 'ностальгию'
 'обязательно' 'одновременно' 'ожидал' 'особенно' 'оставил' 'отличная'
 'отличный' 'оторваться' 'очень' 'первых' 'поворот' 'понравились'
 'понравилось' 'понравился' 'понравится' 'потрясающим' 'предсказуемый'
 'просмотра' 'просмотру' 'работа' 'равно' 'равнодушным' 'режиссера'
 'рекомендую' 'семейного' 'скучновато' 'скучный' 'слабый' 'слишком'
 'смешанные' 'смешно' 'смотрится' 'советую' 'сократить' 'сопровождение'
 'спецэффекты' 'средний' 'сюжет' 'сюжета' 'трогательная' 'трогательно'
 'удивил' 'фантастический' 'фильм' 'фильме

#### 4)	Масштабируйте данные с помощью tf-idf

In [8]:
from sklearn.feature_extraction.text import TfidfVectorizer

stop_words = list(stop_words)
tfidf_vectorizer = TfidfVectorizer(stop_words=stop_words)
X_tfidf = tfidf_vectorizer.fit_transform(reviews)
print(tfidf_vectorizer.get_feature_names_out())

['актерская' 'атмосферу' 'большего' 'вдохновляющий' 'великолепно' 'всем'
 'выглядит' 'вызывает' 'высоте' 'героя' 'главного' 'декорации'
 'депрессивный' 'диалоги' 'длинным' 'дополняло' 'задуматься'
 'замечательная' 'запомнился' 'заставила' 'затянутый' 'затянутыми'
 'захватывает' 'захватывающий' 'игра' 'идеально' 'интересный' 'история'
 'клише' 'конце' 'костюмы' 'легко' 'легкого' 'лучший' 'местами' 'минут'
 'музыкальное' 'невозможно' 'неожиданный' 'ничем' 'ностальгию'
 'обязательно' 'одновременно' 'ожидал' 'особенно' 'оставил' 'отличная'
 'отличный' 'оторваться' 'очень' 'первых' 'поворот' 'понравились'
 'понравилось' 'понравился' 'понравится' 'потрясающим' 'предсказуемый'
 'просмотра' 'просмотру' 'работа' 'равно' 'равнодушным' 'режиссера'
 'рекомендую' 'семейного' 'скучновато' 'скучный' 'слабый' 'слишком'
 'смешанные' 'смешно' 'смотрится' 'советую' 'сократить' 'сопровождение'
 'спецэффекты' 'средний' 'сюжет' 'сюжета' 'трогательная' 'трогательно'
 'удивил' 'фантастический' 'фильм' 'фильме

#### 5)	Использовать n-gram (1, 2, 3), pipeline(vectorizer, regressor)

In [20]:
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
pipeline = Pipeline([
    ('tfidf', TfidfVectorizer(stop_words=stop_words, ngram_range=(1, 4))),
    ('classifier', LogisticRegression()),
])
y = [1, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1]
pipeline.fit(reviews, y)

Pipeline(steps=[('tfidf',
                 TfidfVectorizer(ngram_range=(1, 4),
                                 stop_words=['когда', 'после', 'были', 'ему',
                                             'ведь', 'ты', 'чего', 'тем',
                                             'даже', 'или', 'конечно', 'потом',
                                             'моя', 'тоже', 'только', 'со',
                                             'них', 'вдруг', 'уже', 'об',
                                             'много', 'и', 'их', 'бы', 'что',
                                             'ни', 'был', 'тогда', 'им',
                                             'перед', ...])),
                ('classifier', LogisticRegression())])

#### Оценка модели

In [27]:
new_reviews = ["Невероятный фильм, рекомендую всем!", "Скучный фильм."]
predictions = pipeline.predict(new_reviews)
print("Предсказание:", predictions)
scores = cross_val_score(pipeline, reviews, y, cv=5)
print("Кросс-валидация:", scores)
print(f"Mean cross-validation score: {np.mean(scores):.2f}")
coefficients = pipeline.named_steps['classifier'].coef_
feature_names = pipeline.named_steps['tfidf'].get_feature_names_out()
top_positive_coefficients = np.argsort(coefficients[0])[-10:]
top_negative_coefficients = np.argsort(coefficients[0])[:10]

Предсказание: [1 0]
Кросс-валидация: [0.5        0.83333333 0.8        0.4        0.6       ]
Mean cross-validation score: 0.63


#### 6)	Исследуйте коэффициенты модели

In [28]:
import numpy as np

coefficients = pipeline.named_steps['classifier'].coef_
feature_names = pipeline.named_steps['tfidf'].get_feature_names_out()

top_positive_coefficients = np.argsort(coefficients[0])[-10:]
top_negative_coefficients = np.argsort(coefficients[0])[:10]

print("Позитивные коэффициенты:\n", feature_names[top_positive_coefficients])
print("Негативные коэффициенты:\n", feature_names[top_negative_coefficients])

Позитивные коэффициенты:
 ['удивил очень понравился' 'игра' 'актерская игра' 'актерская' 'просмотру'
 'фильм обязательно просмотру' 'фильм обязательно' 'обязательно'
 'обязательно просмотру' 'очень']
Негативные коэффициенты:
 ['советую' 'оставил' 'фильм оставил' 'скучный' 'скучный фильм'
 'скучный фильм советую' 'фильм советую' 'лучший фильм' 'худший'
 'фильм худший']


#### 7)	Сделайте выводы по работе.

Проанализировали набор отзывов о фильмах. Для анализа использовались 27 отзывов, в которых выражались различные мнения о фильмах - от восторженных до критических.

Был создан мешок слов: собрали отзывы о фильмах и составили набор данных.
Стоп-слова: было создано множество стоп-слов и удалены из отзывов, чтобы сосредоточиться на значимых терминах.
TF-IDF масштабирование: данные были масштабированы с использованием метода TF-IDF, что позволило выделить значимые слова и фразы.
N-gram и обучение модели: спользовали N-gram (1, 2, 3) и настроили конвейер, включающий векторизацию TF-IDF и регрессию. Модель была обучена на отзывов и показала средний кросс-валидационный результат 0.63.
##### Результаты:
Позитивные коэффициенты: фразы, которые чаще всего встречались в положительных отзывах: «удивил очень понравился», «актерская игра», «просмотру».
Негативные коэффициенты: фразы, которые чаще встречались в отрицательных отзывах: «скучный фильм», «фильм оставил», «фильм советую».

Заключение: Анализ отзывов о фильмах с использованием методов обработки естественного языка и машинного обучения позволяет выявить ключевые аспекты, влияющие на восприятие зрителями. Модель показала способность различать положительные и отрицательные отзывы с приемлемой точностью, что может быть полезно для дальнейшего анализа и рекомендаций.